### DISCUSSION TITLE
Adding covariate to joint model

### DISCUSSION THREAD

**Discussion by IJCorreia** [inlabru-org/inlabru#240] [General]
Dear Finn,

I have data set whose rows are detections of animal groups throughout a surveyed region. My intention is to model a marked point process, since I have the group locations, and use the group size as a mark. Below is the code I’m using, following [this](https://github.com/inlabru-org/inlabru/discussions/113) and [this](https://inlabru-org.github.io/inlabru/articles/2d_lgcp_multilikelihood.html) examples. In my case, `data.inlabrutrunc` is the `sf ` and `data.frame` object containing the observed data set, the `tran.inlabru` object (a `sf` and `LINESTRING` object) containing the transects surveyed, and `dem.new.inlabru` is a `SpatRaster` object with seven environmental covariates only for the whole region. For simplicity, I’m only considering fitting univariable models, here with elevation as an example:

```
### ELEVATION
elev <- dem.new.inlabru$elevation #only the layer needed
mean.elev <- mean(terra::values(dem.new.inlabru$elevation), na.rm = TRUE)
elev2 <- elev - mean.elev #centering covariate
elev3 <- elev2
elev3[is.na(elev3)] <- 0 - mean.elev #replacing NAs

par(mfrow = c(1, 3))
plot(elev, main = "elev"); plot(elev2, main = "elev2"); plot(elev3, main = "elev3")
par(mfrow = c(1, 1))
```
![image](https://github.com/inlabru-org/inlabru/assets/50990866/c7e25fd2-ab64-442a-bd0c-81a8d354b61f)

```
ggplot2::ggplot() +
    inlabru::gg(elev3$elevation) + inlabru::gg(meshUTM)
```

![image](https://github.com/inlabru-org/inlabru/assets/50990866/41b097f6-6233-4703-817d-ef702960e29e)

```
### MODEL COMPONENTS
cmp.joint1elev <- ~ Intercept(1) + size.Intercept(1) + sigma_theta(1, prec.linear = 1) +
  mySPDE(main = geometry, model = matern) +
  size.SPDE(main = geometry, model = matern) +
  elev(main = f.elev3(.data.), model = "linear")

### MODEL FORMULAS - HN
fml.joint1elev <- geometry + distance ~ Intercept +
  log(hn(distance, sigma.bru(sigma_theta))) +
  mySPDE + elev + log(2)

#just for group size
fml.jointsize <- size ~ size.Intercept + size.SPDE

### MODEL LIKELIHOOD
lik.1elev <- inlabru::like(family = "cp",
                           formula = fml.joint1elev,
                           data = data.inlabrutrunc,
                           samplers = tran.inlabru,
                           domain = list(geometry = meshUTM,
                                         distance = fmesher::fm_mesh_1d(seq(0, max(bins), length.out = 30))))

#just for group size - negbinom
lik.size <- inlabru::like(family = "nbinomial",
                          formula = fml.jointsize,
                          data = data.inlabrutrunc)#Fitting joint models

### JOINT MODEL
jointfit1elev <- inlabru::bru(cmp.joint1elev, lik.1elev, lik.size,
                              options = list(control.inla = list(int.strategy = "eb"),
                                             #bru_max_iter = 1,
                                             control.predictor = list(compute = TRUE), verbose=TRUE))
```

The model without the covariate runs with no problems but once I run jointfit1elev , after a while, it issues the following error message (with a bit of the verbose in the console, if it is useful):

![image](https://github.com/inlabru-org/inlabru/assets/50990866/e2c6b0fe-24b7-4c7e-9f87-7b5cf9837d8b)

I recently updated R, RStudio and all packages and my session info is presented below. Can you help me understand what I might be doing wrong? I thought about the fact that the covariate raster does not cover all the mesh vertices, but I am not used to the INLA rationale (yet!)... All packages, R and Rstudio are all updated, I checked all units of my data and everything is set to km, the CRS is the same for all spatial objects (for `data.inlabrutrunc`, `tran.inlabru`, and `dem.new.inlabru` -- print below). I also tried centering elevation (as suggested [here](https://inlabru-org.github.io/inlabru/articles/2d_lgcp_covars.html#continuous-covariates) or changing covariates but the error kept showing:

![image](https://github.com/inlabru-org/inlabru/assets/50990866/8fedfcf3-e0ea-4501-8d9e-9842f238f138)

```
> sf::st_crs(data.inlabrutrunc)
Coordinate Reference System:
  User input: PROJCRS["unknown",
    BASEGEOGCRS["WGS 84",
        DATUM["World Geodetic System 1984",
            ELLIPSOID["WGS 84",6378137,298.257223563,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4326]],
    CONVERSION["Polar Stereographic (variant B)",
        METHOD["Polar Stereographic (variant B)",
            ID["EPSG",9829]],
        PARAMETER["Latitude of standard parallel",70,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8832]],
        PARAMETER["Longitude of origin",-45,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8833]],
        PARAMETER["False easting",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["(E)",south,
            MERIDIAN[90,
                ANGLEUNIT["degree",0.0174532925199433,
                    ID["EPSG",9122]]],
            ORDER[1],
            LENGTHUNIT["kilometre",1000]],
        AXIS["(N)",south,
            MERIDIAN[180,
                ANGLEUNIT["degree",0.0174532925199433,
                    ID["EPSG",9122]]],
            ORDER[2],
            LENGTHUNIT["kilometre",1000]]] 
  wkt:
PROJCRS["unknown",
    BASEGEOGCRS["WGS 84",
        DATUM["World Geodetic System 1984",
            ELLIPSOID["WGS 84",6378137,298.257223563,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4326]],
    CONVERSION["Polar Stereographic (variant B)",
        METHOD["Polar Stereographic (variant B)",
            ID["EPSG",9829]],
        PARAMETER["Latitude of standard parallel",70,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8832]],
        PARAMETER["Longitude of origin",-45,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8833]],
        PARAMETER["False easting",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["(E)",south,
            MERIDIAN[90,
                ANGLEUNIT["degree",0.0174532925199433,
                    ID["EPSG",9122]]],
            ORDER[1],
            LENGTHUNIT["kilometre",1000]],
        AXIS["(N)",south,
            MERIDIAN[180,
                ANGLEUNIT["degree",0.0174532925199433,
                    ID["EPSG",9122]]],
            ORDER[2],
            LENGTHUNIT["kilometre",1000]]]
```

```
> sessionInfo()
R version 4.4.0 (2024-04-24 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 22631)

Matrix products: default


locale:
[1] LC_COLLATE=English_United Kingdom.utf8  LC_CTYPE=English_United Kingdom.utf8   
[3] LC_MONETARY=English_United Kingdom.utf8 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.utf8    

time zone: Europe/Lisbon
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] imputeTS_3.3         writexl_1.5.0        stelfi_1.0.1         patchwork_1.2.0      lgcp_2.0            
 [6] inlabru_2.10.1       fmesher_0.1.5        INLA_23.09.09        Matrix_1.7-0         lwgeom_0.2-14       
[11] glue_1.7.0           rlang_1.1.3          rgeos_0.6-4          raster_3.6-26        maptools_1.1-8      
[16] sf_1.0-16            sp_2.1-4             tufte_0.13           RColorBrewer_1.1-3   ggforce_0.4.2       
[21] ggpubr_0.6.0         magrittr_2.0.3       viridis_0.6.5        viridisLite_0.4.2    hrbrthemes_0.8.7    
[26] gridExtra_2.3        tweedie_2.3.5        dsm_2.3.3            numDeriv_2016.8-1.1  mgcv_1.9-1          
[31] nlme_3.1-164         readxl_1.4.3         Distance_1.0.9       mrds_2.3.0           htmlwidgets_1.6.4   
[36] lubridate_1.9.3      forcats_1.0.0        stringr_1.5.1        dplyr_1.1.4          purrr_1.0.2         
[41] readr_2.1.5          tidyr_1.3.1          tibble_3.2.1         ggplot2_3.5.1        tidyverse_2.0.0     
[46] plyr_1.8.9           knitcitations_1.0.12 bibtex_0.5.1         webshot_0.5.5        xtable_1.8-4        
[51] devtools_2.4.5       usethis_2.2.3        bookdown_0.39        knitr_1.47           rmarkdown_2.27      
[56] tinytex_0.51        

loaded via a namespace (and not attached):
  [1] ggtext_0.1.2            fs_1.6.4                spatstat.sparse_3.0-3   httr_1.4.7             
  [5] profvis_0.3.8           tools_4.4.0             backports_1.5.0         utf8_1.2.4             
  [9] R6_2.5.1                rgdal_1.6-7             sn_2.1.1                urlchecker_1.0.1       
 [13] withr_3.0.0             cli_3.6.2               spatstat.explore_3.2-7  labeling_0.4.3         
 [17] tseries_0.10-56         spatstat.data_3.0-4     proxy_0.4-27            systemfonts_1.1.0      
 [21] foreign_0.8-86          gfonts_0.2.0            svglite_2.1.3           sessioninfo_1.2.2      
 [25] maps_3.4.2              TTR_0.24.4              rstudioapi_0.16.0       httpcode_0.3.0         
 [29] generics_0.1.3          spatstat.random_3.2-3   car_3.1-2               fansi_1.0.6            
 [33] RefManageR_1.4.0        abind_1.4-5             terra_1.7-78            lifecycle_1.0.4        
 [37] forecast_8.22.0         yaml_2.3.8              carData_3.0-5           grid_4.4.0             
 [41] promises_1.3.0          crayon_1.5.2            miniUI_0.1.1.1          lattice_0.22-6         
 [45] pillar_1.9.0            tcltk_4.4.0             codetools_0.2-20        wk_0.9.1               
 [49] urca_1.3-4              fontLiberation_0.1.0    data.table_1.15.4       stinepack_1.5          
 [53] remotes_2.5.0           vctrs_0.6.5             spam_2.10-0             cellranger_1.1.0       
 [57] gtable_0.3.5            cachem_1.1.0            xfun_0.44               mime_0.12              
 [61] pracma_2.4.4            rnaturalearth_1.0.1     ncdf4_1.22              timeDate_4032.109      
 [65] sfheaders_0.4.4         iterators_1.0.14        units_0.8-5             fields_15.2            
 [69] statmod_1.5.0           ellipsis_0.3.2          xts_0.13.2              fontquiver_0.2.1       
 [73] Deriv_4.1.3             TMB_1.9.11              KernSmooth_2.23-22      colorspace_2.1-0       
 [77] DBI_1.2.2               nnet_7.3-19             mnormt_2.1.1            tidyselect_1.2.1       
 [81] compiler_4.4.0          extrafontdb_1.0         curl_5.2.1              rvest_1.0.4            
 [85] xml2_1.3.6              fontBitstreamVera_0.1.1 scales_1.3.0            classInt_0.4-10        
 [89] lmtest_0.9-40           fracdiff_1.5-3          quadprog_1.5-8          digest_0.6.35          
 [93] goftest_1.2-3           spatstat.utils_3.0-4    htmltools_0.5.8.1       pkgconfig_2.0.3        
 [97] extrafont_0.19          fastmap_1.2.0           quantmod_0.4.26         shiny_1.8.1.1          
[101] farver_2.1.2            zoo_1.8-12              jsonlite_1.8.8          kableExtra_1.3.4.9000  
[105] s2_1.1.6                dotCall64_1.1-1         munsell_0.5.1           Rcpp_1.0.12            
[109] gdtools_0.3.7           stringi_1.8.4           MASS_7.3-60.2           rpanel_1.1-5.2         
[113] pkgbuild_1.4.4          parallel_4.4.0          deldir_2.0-4            splines_4.4.0          
[117] gridtext_0.1.5          tensor_1.5              hms_1.1.3               spatstat.geom_3.2-9    
[121] ggsignif_0.6.4          stats4_4.4.0            pkgload_1.3.4           crul_1.4.2             
[125] tidyterra_0.6.0         evaluate_0.23           nloptr_2.0.3            tzdb_0.4.0             
[129] tweenr_2.0.3            httpuv_1.6.15           MatrixModels_0.5-3      Rttf2pt1_1.3.12        
[133] polyclip_1.10-6         broom_1.0.6             e1071_1.7-14            Rsolnp_1.16            
[137] rstatix_0.7.2           later_1.3.2             class_7.3-22            intervals_0.15.4       
[141] truncnorm_1.0-9         optimx_2023-10.21       memoise_2.0.1           timechange_0.3.0
```

**Comment by IJCorreia** [comment]
Also, forgot to add:

```
f.elev3 <- function(where) {
  # Extract the values
  v <- inlabru::eval_spatial(elev3$elevation, where, layer = "elevation")
  # Fill in missing values
  # if (any(is.na(v))) {
  #   v <- inlabru::bru_fill_missing(elev3$elevation, where, v)
  # }
  # return(v)
}
```

**Reply by finnlindgren** [reply]
I would treat " I thought about the fact that the covariate raster does not cover all the mesh vertices, but I am not used to the INLA rationale (yet!)..." as the likely problem. You _cannot_ have missing covariate values in the estimation domain, or just outside it (as the integration points may like slightly outside the integration domain if it doesn't align perfectly with the mesh triangle edges).  Note the missing calls to bru_fill_missing in your `f.eval3` code. I would recommend adding a step to fill in missing values in the covariate before doing the estimation. That way, you can more easily check that it's filling in with sensible values (they only need to be sensible in the vicinity of the integration domain, i.e. the "samplers"). Filling in the missing values inside the .eval function was an early "quick and dirty" workaround that I would now recommend against, at it hides potential problems with the data.

**Reply by IJCorreia** [reply]
Thank you for your reply! I noticed I provided `f.elev3` mostly all commented, but that was a typo, I ran my code with the `return(v)` uncommented, but after that tried without commenting the `bru_fill_missing` part. Anyway, this time, I ran my code with `f.elev3` as

```
f.elev3 <- function(where) {
  # Extract the values
  v <- inlabru::eval_spatial(elev3$elevation, where, layer = "elevation")
  # Fill in missing values
  if (any(is.na(v))) {
    v <- inlabru::bru_fill_missing(elev3$elevation, where, v)
  }
  return(v)
}
```

Further, I managed to have access to a larger version of the elevation data, covering everything (points - black, samplers - white, polygons - red, and mesh), I guess it should be enough, right? (below)

![image](https://github.com/inlabru-org/inlabru/assets/50990866/2f8d1491-9234-4b09-a744-ef02d44d3dc0)

I tried fitting the same model but still the session aborts...

![image](https://github.com/inlabru-org/inlabru/assets/50990866/b62bdd0e-f98d-4c7d-9c00-39ce4e641abf)

In the meantime, I was wondering if the `eval_spatial` function runs on the mesh too? I am asking this because even though the CRS for all the objects presented are the same, when I run `f.elev3(meshUTM$loc)` it returns a vector solely made of NAs (probably what is causing the problem here?) and, by plotting a similar `ggplot` as above, I get:

```
ggplot2::ggplot() +
  inlabru::gg(elev3$elevation) + 
  inlabru::gg(meshUTM$loc) + #here is the difference
  inlabru::gg(border.utm, alpha = 0.33, color = "red") +
  inlabru::gg(tran.inlabru, color = "white") + 
  inlabru::gg(data.inlabrutrunc, alpha = 0.5, color = "black", size = 0.5)
```

![image](https://github.com/inlabru-org/inlabru/assets/50990866/97894394-91d9-4202-a621-b01210273328)

P.S.: I also tried using the exact same procedure but for another variable (slope, below) and the following error appeared (with a bit of verbose in the console, if it is useful):

![image](https://github.com/inlabru-org/inlabru/assets/50990866/ef39b304-7242-476f-bde7-1b39127a9e1c)

Thank you once again for you help and patience! I will be waiting for your feedback!

**Reply by finnlindgren** [reply]
The mesh nodes are stored as a raw coordinate matrix, so giving that to gg bypasses the CRS information in the mesh. Try `gg(meshUTM)`. Although if the other objects are in the same coordinate system, that shouldn't make a difference, so there appears to be something wrong with your crs specifications. When the covariate and mesh have the same crs, `eval_spatial(elev3$elevation, meshUTM$loc)` should not return any NAs.

Please check `fm_crs(object)` for all of the involved spatial objects, and the mesh, to verify that they are indeed identical.

**Reply by finnlindgren** [reply]
The "x values must be strictly increasing" might be linked to using a rw1/2 model with invalid specification, but in your case I'm not sure, as you first get the NAN/INF warnings.

**Reply by IJCorreia** [reply]
Ok, I understand and tried what you suggested. Here are the two plots:

```
plot1 <- ggplot2::ggplot() +
  inlabru::gg(elev3$elevation) +
  inlabru::gg(meshUTM) +
  inlabru::gg(border.utm, alpha = 0.33, color = "red") +
  inlabru::gg(tran.inlabru, color = "white") +
  inlabru::gg(data.inlabrutrunc, alpha = 0.5, color = "black", size = 0.5)

plot2 <- ggplot2::ggplot() +
  inlabru::gg(elev3$elevation) +
  inlabru::gg(meshUTM$loc) + #here is the difference
  inlabru::gg(border.utm, alpha = 0.33, color = "red") +
  inlabru::gg(tran.inlabru, color = "white") +
  inlabru::gg(data.inlabrutrunc, alpha = 0.5, color = "black", size = 0.5)

gridExtra::grid.arrange(plot1, plot2, ncol = 2)
```

![image](https://github.com/inlabru-org/inlabru/assets/50990866/e1f44876-f105-4fa9-b7e8-22e3748a3a7e)

I also understand that it may be related with the CRS specification, but this is a bit confusing for me to detect where it is coming from... So, concerning the `fm_crs(object)` part you asked, the output is quite long but I tried to make it a bit more organized in the attached .pdf, perhaps it is better:

[fm_crs_Finn.pdf](https://github.com/user-attachments/files/15986735/fm_crs_Finn.pdf)

And they seem to be similar...

![image](https://github.com/inlabru-org/inlabru/assets/50990866/8d00c3f6-bd08-48df-9f44-eaaf766d33c2)

Except for the `meshUTM$loc` CRS, should it be changed to the same CRS? Shouldn't this CRS for the `meshUTM$loc` be automatically "herited" when I create the mesh? 
Below is the code I use to create the mesh, just in case:

```
meshUTM <- fmesher::fm_mesh_2d(loc.domain = sf::st_as_sfc(border.utm), max.edge = c(25, 70)); plot(meshUTM)
fmesher::fm_crs(meshUTM) <- fmesher::fm_crs(sf::st_crs(sf::st_as_sfc(border.utm)))
```

Finally, about your last comment, if your guess is correct, where can I check the RW1/2 model with specification?
Thank you, yet again!

**Reply by finnlindgren** [reply]
The `$loc` data is stored as a plain matrix, so doesn't have a crs; only the mesh itself knows what CRS those raw coordinates are stored in. You need to check if
```
fmesher::fm_crs(elev3$elevation) == fmesher::fm_crs(meshUTM)
```
returns `TRUE`. If it doesn't, you need to convert the mesh to have matching crs.

For rw1/2, your model specification above doesn't include any such components, so that can't be the issue.

**Reply by finnlindgren** [reply]
You shouldn't forcibly modify the mesh crs (I know some tutorials out there do that, but that is unsupported and can give an invalid/incorrect mesh). Instead, do
```
meshUTM <- fmesher::fm_mesh_2d(
  loc.domain = sf::st_as_sfc(border.utm),
  max.edge = c(25, 70),
  crs = sf::st_crs(sf::st_as_sfc(border.utm)))
plot(meshUTM)
```

**Reply by finnlindgren** [reply]
From the pdf, everything looks fine. The plot anomaly might just be a quirk in how ggplot handles combinations of stereographic coordinates and raw non-CRS coordinates. The plot with the mesh looks fine.

One potential issue I didn't spot before is that the elevation is given on a much finer resolution than the mesh. This means that the default integration scheme, which is based just on the mesh resolution, may be insufficient, and you need to use a finer integration scheme for the point process likelihood. If you install the development version of `fmesher`, you can use `geometry = fm_subdivide(meshUTM, n = 3)` to increase the integration scheme resolution in the domain specification.

**Reply by IJCorreia** [reply]
Thank you for all your help, Finn, it worked! 
I updated the mesh code as you suggested and it seems that the `geometry = fm_subdivide(meshUTM, n = 3)` did the trick for me to be able to incorporate environmental covariates into the model!
Since this seems to be related with the resolution, does this mean the raster resolution was kind of "overwhelming" in the integration scheme? Can you please explain a bit more about this detail?

**Reply by finnlindgren** [reply]
In short, if a covariate varies at a smaller level of detail than the integration scheme, and (sufficiently many of) the observed points are located at places where the covariate is large (in a way that's missed by the integration scheme, by falling somewhere between the integration points), then the likelihood approximation isn't good. this is because the point process log-likelihood has one term _penalising_ the total intensity integral, and another term that _rewards_ large intensity. When the integration scheme is too coarse, those rewards aren't properly balanced by the integral penalty. We're currently writing a paper showing how this works in practice.
